File: manual/linear_model.py
import sys
import pandas as pd
import matplotlib.pyplot as plt
from scipy.stats import linregress
from sklearn.metrics import mean_squared_error

if len(sys.argv) != 4:
    print("Usage: python linear_model.py <filename> <x_column> <y_column>")
    sys.exit(1)

filename = sys.argv[1]
x_col = sys.argv[2]
y_col = sys.argv[3]

try:
    data = pd.read_csv(filename)
except FileNotFoundError:
    print(f"Error: File '{filename}' not found.")
    sys.exit(1)
x = data[x_col]
y = data[y_col]

slope, intercept, r_value, p_value, std_err = linregress(x, y)
y_pred = slope * x + intercept
mse = mean_squared_error(y, y_pred)

print(f"Slope: {slope:.4f}")
print(f"Intercept: {intercept:.4f}")
print(f"Correlation coefficient (r): {r_value:.4f}")
print(f"MSE: {mse:.4f}")

plt.scatter(x, y, color="red")
plt.plot(x, y_pred, color="blue")
label = "y = {:.2f}x + {:.2f}\\nr = {:.2f}\\nMSE = {:.2f}".format(slope, intercept, r_value, mse)
plt.text(0.05, 0.95, label,
         transform=plt.gca().transAxes, fontsize=10, verticalalignment="top")
plt.xlabel(x_col)
plt.ylabel(y_col)
plt.title(f"{y_col} vs {x_col}")
plt.savefig("regression_plot_python.png")
plt.show()

File: manual/linear_model.R
args <- commandArgs(trailingOnly = TRUE)
if (length(args) != 3) {
  stop("Usage: Rscript linear_model.R <filename> <x_column> <y_column>")
}
filename <- args[1]
x_col <- args[2]
y_col <- args[3]
data <- read.csv(filename)
x <- data[[x_col]]
y <- data[[y_col]]
model <- lm(y ~ x)
slope <- coef(model)[2]
intercept <- coef(model)[1]
r <- cor(x, y)
pred <- predict(model)
mse <- mean((y - pred)^2)
cat("Slope:", round(slope, 4), "\n")
cat("Intercept:", round(intercept, 4), "\n")
cat("Correlation coefficient (r):", round(r, 4), "\n")
cat("MSE:", round(mse, 4), "\n")
library(ggplot2)
p <- ggplot(data, aes_string(x = x_col, y = y_col)) +
  geom_point(color = "red") +
  geom_smooth(method = "lm", se = FALSE, color = "blue") +
  annotate("text", x = min(x) + 0.1 * (max(x) - min(x)), y = max(y) * 0.95,
           label = paste("y =", round(slope, 2), "x +", round(intercept, 2), "\nr =", round(r, 2), "\nMSE =", round(mse, 2)),
           hjust = 0, size = 4) +
  labs(title = paste(y_col, "vs", x_col), x = x_col, y = y_col) +
  theme_minimal()
ggsave("regression_plot_r.png", p, width = 7, height = 5)
print(p)
